pkg/localstorev2/reserve.go

```diff
@@ -0,0 +1,496 @@
+// Copyright 2023 The Swarm Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package storer
+
+import (
+	"bytes"
+	"context"
+	"crypto/hmac"
+	"encoding/binary"
+	"encoding/hex"
+	"errors"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/ethersphere/bee/pkg/bmtpool"
+	"github.com/ethersphere/bee/pkg/postage"
+	storage "github.com/ethersphere/bee/pkg/storagev2"
+	"github.com/ethersphere/bee/pkg/swarm"
+	"go.uber.org/atomic"
+	"golang.org/x/sync/errgroup"
+)
+
+const (
+	reserveOverCapacity = "reserveOverCapacity"
+	reserveUnreserved   = "reserveUnreserved"
+	reserveLock         = "reserveLock"
+)
+
+var errMaxRadius = errors.New("max radius reached")
+
+type SyncReporter interface {
+	// Number of active historical syncing jobs.
+	Rate() float64
+}
+
+func (db *DB) reserveWorker(capacity int, syncer SyncReporter, warmupDur, wakeUpDur time.Duration) {
+	defer db.reserveWg.Done()
+
+	threshold := capacity * 4 / 10
+
+	overCapTrigger, overCapUnsub := db.events.Subscribe(reserveOverCapacity)
+	defer overCapUnsub()
+
+	select {
+	case <-time.After(warmupDur):
+	case <-db.quit:
+		return
+	}
+
+	ctx, cancel := context.WithCancel(context.Background())
+	go func() {
+		<-db.quit
+		cancel()
+	}()
+
+	for {
+		select {
+		case <-overCapTrigger:
+			err := db.unreserve(ctx)
+			if err != nil {
+				db.logger.Error(err, "reserve unreserve")
+			}
+		case <-time.After(wakeUpDur):
+			radius := db.reserve.Radius()
+			if db.reserve.Size() < threshold && syncer.Rate() == 0 && radius > 0 {
+				radius--
+				err := db.reserve.SetRadius(db.repo.IndexStore(), radius)
+				if err != nil {
+					db.logger.Error(err, "reserve set radius")
+				}
+				db.logger.Info("reserve radius decrease", "radius", radius)
+
+			}
+		case <-db.quit:
+			return
+		}
+	}
+}
+
+func (db *DB) ReserveGet(ctx context.Context, addr swarm.Address, batchID []byte) (swarm.Chunk, error) {
+	return db.reserve.Get(ctx, db.repo, addr, batchID)
+}
+
+func (db *DB) ReserveHas(addr swarm.Address, batchID []byte) (bool, error) {
+	return db.reserve.Has(db.repo.IndexStore(), addr, batchID)
+}
+
+// ReservePutter returns a PutterSession for inserting chunks into the reserve.
+func (db *DB) ReservePutter(ctx context.Context) PutterSession {
+
+	trx, commit, rollback := db.repo.NewTx(ctx)
+
+	triggerBins := make(map[uint8]bool)
+	count := 0
+
+	// lock to avoid Putting a chunk that expires during the session.
+	db.lock.Lock(reserveLock)
```

Comment:
This could be okay I think, but would it be better to have the Lock/Unlock in the individual functions of the putterSession? This way we will have lesser lock contention. Multiple calls to ReservePutter can also go through. For eg lot of routines would be using this function simultaneously in pullsync I think. Wdyt?

Reply:
do you mean lock inside `PutterFunc` ?